nio.json

```diff
@@ -0,0 +1,162 @@
+            "jwt_secret": "",
+            "ip_address": "",
+            "interface": "",
+            "port": "",
+            "connect_timeout": 2,
+            "websocket_ping_interval": 10,
+            "use_localhost": false
+          },
+          "websocket": {
+            "ws_host": "",
+            "ws_port": "",
+            "ws_secure": ""
+          },
+          "pubkeeper.brew.zmq.brew.ZMQBrew": {
+            "ip_address": "",
+            "publisher_min_port": 9010,
+            "publisher_max_port": 9199,
+            "linger": 1000,
+            "hwm_size": 100000,
+            "max_sockets": 1023,
+            "max_io_thread_count": 1,
+            "encryption": {
+              "allowed_ip_addresses": "",
+              "publisher_public_keys_dir": "",
+              "publisher_server_public_cert": "",
+              "publisher_server_secret_cert": "",
+              "subscriber_client_secret_cert": ""
+            }
+          }
+        }
+      },
+      "scheduler": {
+        "custom": {
+          "min_interval": 0.1,
+          "resolution": 0.1
+        }
+      },
+      "security": {
+        "basic": {
+          "users": "etc/users.json",
+          "permissions": "etc/permissions.json"
+        },
+        "oauth2": {
+          "permissions": "etc/permissions.json",
+          "client_id": "407408718192.apps.googleusercontent.com",
+          "validate_url": "https=/www.googleapis.com/oauth2/v2/tokeninfo?access_token=%s"
+        },
+        "jwt": {
+        }
+      },
+      "web": {
+        "cherrypy": {
+          "server_thread_pool": 50,
+          "request_show_tracebacks": false,
+          "cors_allow_origin": "*"
+        }
+      }
+    },
+    "components": {
+      "snmp": {
+        "host": "127.0.0.1",
+        "port": 161,
+        "community_index": "agent",
+        "community_name": "public",
+        "trap_host": "127.0.0.1",
+        "trap_port": 162,
+        "trap_community_index": "agent",
+        "status_traps": "error"
+      },
+      "management_publisher": {
+        "topic": "management",
+        "add_instance_id": true
+      },
+      "rest": {
+        "port": "",
+        "host": "",
+        "network_retry_interval": 30,
+        "ssl_certificate": "",
+        "ssl_private_key": "",
+        "ssl_certificate_chain": ""
+      },
+      "diagnostic": {
+        "api_url_prefix": "https://api.nio.works/v1/diagnostic",
+        "service": true,
+        "signals": true,
+        "failure_queue": true,
+        "failure_queue_length": 10000,
+        "add_instance_id": true
+      },
+      "service_monitor": {
+        "healthy_monitor_interval": 30,
+        "failed_monitor_interval": 60,
+        "restart_service_interval": -1
+      },
+      "log_api": {
+      },
+      "config": {
+      },
+      "project_manager": {
+      }
+    }
+  },
+  "user_defined": {
+    "TWITTER_API_KEY": "EXAMPLE-as39fhy283hfa38fha",
+    "NIO_HOST": "127.0.0.1",
```

Comment:
This feels wrong.  Much like using env-variables for things.  This is an actual system setting, that happens to be user-configurable.  I'd stick it where it ought to be, in the rest component?

Or, at least have a syntax like `[[NIO_HOST]]` within the `system.component.rest.host` element that will look to a user-defined variable at least.

Reply:
good call